aquifer/aquifer
# ES6, Promises, and general cleanup

This PR introduces some changes:
- All code has been updated to make use of latest ES6 sugar.
- All API libraries are now JavaScript classes.
- Updates the Aquifer app delegate so that it is cleaner and faster.
- Code now uses ES6 Promises everywhere to ensure that sync/async is 100% controllable at all times.
- Use of Promises enables much more consistent and correct error messages.
- Updates console API so that it now returns proper error codes, and uses a better color for info messages.
- Cleans up command files, and loads them more efficiently.
- Removes drush lock file compatibility from the build system. Nobody was using that feature, because it doesn't work consistently. Instead, we should opt to explicitly define versions in our make files.
- Fixes some issues with the Extension system.
  - `Please run "aquifer extensions-load"` no longer appears while you're running `aquifer extensions-load`
  - All child processes now pipe stdout and stderr to the parent stdout and stderr streams.
## API change

Each class, although attached and available through an Aquifer instance, is not aware of a specific Aquifer instance, and so it must be passed in, like so:

``` javascript
// Create an instance of the build class.
let Build = new Aquifer.build(Aquifer);

// Set up parameters needed to create a build.
let destination = '/var/www/project/build';
let options = {
  symlink: true,
  makeFile: `/var/www/project/drush.make.yml`
};

// Create build, and make sure error/success messages are returned.
Build.create(destination, options)
.then(() => {
  Aquifer.console.log('Awesome, the build finished!');
})
.catch((reason) => {
  // Log with error code '1'.
  Aquifer.console.log('The build failed for the following reason: ' + reason, 'error', 1);
});
```

This gives us more control over the context in which APIs are executed, this will become even more useful when we run tests.
## Resolves:

The structural changes made in this update resolve several of the issues we have:
- https://github.com/aquifer/aquifer/issues/90
- https://github.com/aquifer/aquifer/issues/95
- https://github.com/aquifer/aquifer/issues/55
## Steps to test
- Checkout this branch of Aquifer.
- Run `aquifer` and ensure that you see the normal documentation.
- Run `aquifer create test` and ensure that Aquifer creates a full project as expected.
- `cd test`
- Run `aquifer build --help` and ensure that you see no documentation on using lock files.
- Run `aquifer build` and ensure that a full build is created with the correct symlinks.
- Run `aquifer build -c` and ensure Aquifer rebuilds, and copies custom code instead of symlinking it.
- Run `aquifer aquifer build -d /path/to/a/build/dest` and ensure that Aquifer correctly builds into that directory.
- Run `aquifer extension-add aquifer-coder`
- Ensure that Aquifer thinks for a minute, then returns NPM's stdout, and gives you a success message.
- Run `aquifer extensions-list` and ensure `aquifer-coder` is returned.
- Run `rm -r .aquifer/node_modules/`
- Run `aquifer` and ensure it gives you a message about `aquifer-coder` not being loaded.
- Run `aquifer extensions-load`
  - Ensure that it does not deliver the "missing extension" message
  - Ensure that it re-downloads `aquifer-coder`
- Run `aquifer extension-remove aquifer-coder` and ensure that it removes aquifer-coder from your project.
- Edit the `drush.make.yml` file, and create a syntax error.
- Run `aquifer build` and ensure that it returns an error with a stack trace and exits with the code 1.
- In an Aquifer-based project that has real code and an environment, run `aquifer refresh -a @alias`
- Ensure that Aquifer runs all refresh commands in the correct order successfully.
